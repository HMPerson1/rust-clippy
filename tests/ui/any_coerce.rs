// Copyright 2014-2018 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

#![feature(unsize, coerce_unsized)]
#![deny(clippy::wrong_any_coerce)]
#![deny(bare_trait_objects)]

use std::any::Any;
use std::cell::RefCell;
use std::fmt::Debug;
use std::iter::Iterator;
use std::marker::{Send, Unsize};
use std::ops::CoerceUnsized;
use std::ops::Deref;
use std::rc::Rc;

struct Foo;

struct Unsizeable<T: ?Sized, U: ?Sized, V: ?Sized> {
    box_v: Box<V>,
    rc_t: Rc<T>,
    u: U,
}

fn main() {
    let mut box_any: Box<dyn Any + Send> = Box::new(Foo);
    let _: *mut dyn Any = &mut box_any; // LINT
    let _: *mut dyn Any = &mut *box_any; // ok

    let rc_rc_any: Rc<Rc<dyn Any>> = Rc::new(Rc::new(Foo));
    let _: &dyn Any = &rc_rc_any; // LINT
    let _: &dyn Any = &*rc_rc_any; // LINT
    let _: &dyn Any = &**rc_rc_any; // ok
    let _: &Rc<dyn Any> = &*rc_rc_any; // ok

    let refcell_box_any: RefCell<Box<dyn Any>> = RefCell::new(Box::new(Foo));
    let _: &RefCell<dyn Any> = &refcell_box_any; // LINT

    let rc_unsizable_rc_any: Rc<Unsizeable<i32, Rc<dyn Any>, i32>> = Rc::new(Unsizeable {
        box_v: Box::new(0),
        rc_t: Rc::new(0),
        u: Rc::new(Foo),
    });
    let _: Rc<Unsizeable<i32, dyn Any, i32>> = rc_unsizable_rc_any.clone(); // LINT
    let _: &Unsizeable<i32, dyn Any, i32> = &*rc_unsizable_rc_any; // LINT
    let _: &Rc<Unsizeable<i32, Rc<dyn Any>, i32>> = &rc_unsizable_rc_any; // ok
    let _: &Unsizeable<i32, Rc<dyn Any>, i32> = &*rc_unsizable_rc_any; // ok

    let ref_any: &dyn Any = &Foo;
    let _: &dyn Any = &ref_any; // LINT
    let _: &dyn Any = &*ref_any; // ok

    let ref_refcell_any: &'static RefCell<dyn Any> = Box::leak(Box::new(RefCell::new(Foo)));
    let _: &dyn Any = &ref_refcell_any.borrow(); // LINT
    let _: &dyn Any = &*ref_refcell_any.borrow(); // ok
}

fn very_generic<T, U>(t: &'static T)
where
    T: Deref<Target = U> + 'static,
    U: Deref<Target = dyn Any + Send> + 'static,
{
    let _: &dyn Any = t; // LINT
    let _: &dyn Any = &t; // LINT
    let _: &dyn Any = &*t; // LINT
    let _: &dyn Any = &**t; // LINT
    let _: &dyn Any = &***t; // ok
}
