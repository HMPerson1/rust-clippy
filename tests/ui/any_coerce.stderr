error: coercing `std::boxed::Box<(dyn std::any::Any + std::marker::Send + 'static)>` to `dyn Any`
  --> $DIR/any_coerce.rs:33:27
   |
LL |     let _: *mut dyn Any = &mut box_any; // LINT
   |                           ^^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/any_coerce.rs:11:9
   |
LL | #![deny(clippy::wrong_any_coerce)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^

error: coercing `std::rc::Rc<std::rc::Rc<(dyn std::any::Any + 'static)>>` to `dyn Any`
  --> $DIR/any_coerce.rs:37:23
   |
LL |     let _: &dyn Any = &rc_rc_any; // LINT
   |                       ^^^^^^^^^^

error: coercing `std::rc::Rc<(dyn std::any::Any + 'static)>` to `dyn Any`
  --> $DIR/any_coerce.rs:38:23
   |
LL |     let _: &dyn Any = &*rc_rc_any; // LINT
   |                       ^^^^^^^^^^^

error: coercing `std::boxed::Box<(dyn std::any::Any + 'static)>` to `dyn Any`
  --> $DIR/any_coerce.rs:43:32
   |
LL |     let _: &RefCell<dyn Any> = &refcell_box_any; // LINT
   |                                ^^^^^^^^^^^^^^^^

error: coercing `std::rc::Rc<(dyn std::any::Any + 'static)>` to `dyn Any`
  --> $DIR/any_coerce.rs:50:48
   |
LL |     let _: Rc<Unsizeable<i32, dyn Any, i32>> = rc_unsizable_rc_any.clone(); // LINT
   |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: coercing `std::rc::Rc<(dyn std::any::Any + 'static)>` to `dyn Any`
  --> $DIR/any_coerce.rs:51:45
   |
LL |     let _: &Unsizeable<i32, dyn Any, i32> = &*rc_unsizable_rc_any; // LINT
   |                                             ^^^^^^^^^^^^^^^^^^^^^

error: coercing `&'static (dyn std::any::Any + 'static)` to `dyn Any`
  --> $DIR/any_coerce.rs:56:23
   |
LL |     let _: &dyn Any = &ref_any; // LINT
   |                       ^^^^^^^^

error: coercing `std::cell::Ref<'static, (dyn std::any::Any + 'static)>` to `dyn Any`
  --> $DIR/any_coerce.rs:60:23
   |
LL |     let _: &dyn Any = &ref_refcell_any.borrow(); // LINT
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^

error: coercing `T` to `dyn Any`
  --> $DIR/any_coerce.rs:69:23
   |
LL |     let _: &dyn Any = t; // LINT
   |                       ^
   |
   = note: `T` dereferences to `dyn Any`

error: coercing `&'static T` to `dyn Any`
  --> $DIR/any_coerce.rs:70:23
   |
LL |     let _: &dyn Any = &t; // LINT
   |                       ^^
   |
   = note: `&'static T` dereferences to `dyn Any`

error: coercing `T` to `dyn Any`
  --> $DIR/any_coerce.rs:71:23
   |
LL |     let _: &dyn Any = &*t; // LINT
   |                       ^^^
   |
   = note: `T` dereferences to `dyn Any`

error: coercing `U` to `dyn Any`
  --> $DIR/any_coerce.rs:72:23
   |
LL |     let _: &dyn Any = &**t; // LINT
   |                       ^^^^
   |
   = note: `U` dereferences to `dyn Any`

error: aborting due to 12 previous errors

